llvm/lib/CAS/BuiltinCAS.cpp

```diff
@@ -2704,7 +2704,7 @@ OnDiskCAS::getObjectProxy(IndexProxy I) const {
   SmallString<256> Path;
   getStandalonePath(Object.SK, I, Path);
   ErrorOr<std::unique_ptr<MemoryBuffer>> OwnedBuffer = MemoryBuffer::getFile(
-      Path, /*IsText=*/false, /*RequiresNullTerminator=*/false);
+      Path, /*IsText=*/false, /*RequiresNullTerminator=*/true);
```

Comment:
This isn't the right fix. The logic in the OnDiskCAS should guarantee null-termination (somehow) so we can use the `mmap` instead of doing a `memcpy`. This is the point of having a `Blob0` file format.

The bug must be somewhere else.

Reply:
I've looked over the logic in this function a few times. I *think* it's correct, so it must be on the record creation size.

It's possible this is filesystem corruption, but I kind of doubt it. Or I may not have read this code correctly and there's a bug here.

Assume the problem is in creation, it's important to know what the `TrieRecord::StorageKind` is when the assertion fails:
- If it's `TrieRecord::Blob0`, the problem is that we failed to actually append a 0.
- If it's `TrieRecord::Blob`, the problem is that we have a bad page size calculation. We expected the size NOT to be aligned to the page size, which makes the `mmap` work, but then in practice it ended up being aligned to page size.
- If it's `TrieRecord::Data`, the problem is that we didn't 0-pad the data correctly.

Reply:
Note that if it is data corruption, the CAS instance will permanently be poisoned because the bad object will be in there (and found again every time you run).

Probably worth ruling that out:
- Use a different/clean/fresh CAS instance. If the problem still reproduces, it's probably not data corruption, unless it's a concurrency issue that causes the corruption.
- Use a different/clean/fresh CAS instance AND turn off the multithreading in llvm-cas-object-format (I think `--num-threads=1`?). If the problem still reproduces, it's not a concurrency issue that causes the corruption. Otherwise, if the previous one did reproduce, it probably is a concurrency issue.

Reply:
Hi @dexonsmith , I tried to create a new cas instance and it crashed, I then tried to create a new cas instance and used --num-threads=1 (that is the correct option) and it still crashed at the same place. Do you want me to file a radar for this?

Reply:
Great, so that rules out concurrency / multithreading.

> Assume the problem is in creation, it's important to know what the TrieRecord::StorageKind is when the assertion fails:

Have you been able to open this up in a debugger to determine this?

Reply:
> I am not sure what you mean when you say:
> 
> > If it's TrieRecord::Blob0, the problem is that we failed to actually append a 0.
> > If it's TrieRecord::Blob, the problem is that we have a bad page size calculation. We expected the size NOT to be aligned to the page size, which makes the mmap work, but then in practice it ended up being aligned to page size.
> > If it's TrieRecord::Data, the problem is that we didn't 0-pad the data correctly.
> 
> The StorageKind is `Standalone`
> 
> ```
> (lldb) p Object.SK
> (TrieRecord::StorageKind) $1 = Standalone
> ```

Thanks, I had "data" from the file extension; `Standalone` is what I meant by `Data`.